introduction_to_amazon_algorithms/xgboost_abalone/xgboost_abalone.ipynb

```diff
@@ -563,15 +478,138 @@
    "source": [
     "client.delete_endpoint(EndpointName=endpoint_name)"
    ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Appendix"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Data split and upload\n",
+    "\n",
+    "Following methods split the data into train/test/validation datasets and upload files to S3."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import io\n",
+    "import boto3\n",
+    "import random\n",
+    "\n",
+    "\n",
+    "def data_split(\n",
```

Comment:
Same comment as on other notebook. This function seems unnecessarily complicated. If there's a built-in function we can use instead let's prefer that, otherwise I would refactor as 


```python
def data_split(
    FILE_DATA, FILE_TRAIN, FILE_VALIDATION, FILE_TEST, PERCENT_TRAIN, PERCENT_VALIDATION, PERCENT_TEST
):
    
    with open(FILE_DATA, "r") as f:
         data = f.readlines()
    num_data_points = len(data)

    num_train = int((PERCENT_TRAIN / 100.0) * num_data_points)
    num_valid = int((PERCENT_VALIDATION / 100.0) * num_data_points)
    num_tests = int((PERCENT_TEST / 100.0) * num_data_points)

    data_fractions = [num_train, num_valid, num_tests]
    split_data = [[], [], []]

    rand_data_ind = 0

    for split_ind, fraction in enumerate(data_fractions):
        for i in range(fraction):
            rand_data_ind = random.randint(0, len(data) - 1)
            split_data[split_ind].append(data[rand_data_ind])
            data.pop(rand_data_ind)

    with open(FILE_TRAIN, "w") as train_file:
        for l in split_data[0]:
            train_file.write(l)
    with open(FILE_VALIDATION, "w") as valid_file:
        for l in split_data[1]:
            valid_file.write(l)
    with open(FILE_TEST, "w") as tests_file:
        for l in split_data[2]:
            tests_file.write(l)

```

Mutating `data` by popping points in the first loop is pretty inefficient. It would be better to construct a map of indices perhaps, but maybe it isn't necessary as long as the dataset is small.

Reply:
As I mentioned in the other notebook, this is the same code that was written earlier. I just moved it from top to the Appendix. Better to not change it at the last minute unless you strongly feel otherwise.